### PR TITLE
HADOOP-15847 limit the r/w capacity 

### PR DESCRIPTION
**Pull request by lqjack**

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 31 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1058 | trunk passed |
| +1 | compile | 36 | trunk passed |
| +1 | checkstyle | 24 | trunk passed |
| +1 | mvnsite | 41 | trunk passed |
| +1 | shadedclient | 745 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | trunk passed |
| 0 | spotbugs | 59 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 57 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 32 | the patch passed |
| +1 | compile | 30 | the patch passed |
| +1 | javac | 30 | the patch passed |
| -0 | checkstyle | 19 | hadoop-tools/hadoop-aws: The patch generated 12 new + 0 unchanged - 0 fixed = 12 total (was 0) |
| +1 | mvnsite | 34 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | xml | 2 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 755 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 24 | the patch passed |
| +1 | findbugs | 63 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 276 | hadoop-aws in the patch passed. |
| +1 | asflicense | 33 | The patch does not generate ASF License warnings. |
| | | 3375 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1037 |
| Optional Tests | dupname asflicense mvnsite compile javac javadoc mvninstall unit shadedclient findbugs checkstyle xml |
| uname | Linux 633be8e42841 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / d203045 |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/1/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/1/testReport/ |
| Max. process+thread count | 444 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestS3GuardConcurrentOps.java [hunk 1]
```diff
@@ -55,6 +57,23 @@
...
+    Configuration conf =  super.createConfiguration();
+    //patch the read/write capacity
+    boolean scaleCapacityLimitEnabled = conf.getBoolean("fs.s3a.s3guard.ddb.table.scale.capacity.limit", true);
+    if(scaleCapacityLimitEnabled) {
+      LOG.info("Enable the capacity limit : {} -> {}, {} -> {}",
+        S3GUARD_DDB_TABLE_CAPACITY_READ_KEY, 1, S3GUARD_DDB_TABLE_CAPACITY_WRITE_KEY,1);
+      conf.set(S3GUARD_DDB_TABLE_CAPACITY_READ_KEY, "1");
+      conf.set(S3GUARD_DDB_TABLE_CAPACITY_WRITE_KEY, "1");
+    }
+    if(scaleCapacityLimitEnabled) {
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestS3GuardConcurrentOps.java]: we don't need these -they don't show much.

What would be better would be the thread which actually creates a table asking for the bucket metadata and working out the capacity there. But I'm happy not to worry about that...All I'd prefer is to make sure that we don't run up any bills from test tables which weren't deleted

File: hadoop-tools/hadoop-aws/src/test/resources/core-site.xml [hunk 2]
```diff
@@ -153,11 +153,11 @@
   <!-- Reduce DDB capacity on auto-created tables, to keep bills down. -->
   <property>
     <name>fs.s3a.s3guard.ddb.table.capacity.read</name>
-    <value>10</value>
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/resources/core-site.xml]: wow, I'd missed completely we were setting values here

These need to be changed to 0, to ensure the tables are always on demand

File: hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestS3GuardConcurrentOps.java [hunk 3]
```diff
@@ -55,6 +57,23 @@
   @Rule
   public final Timeout timeout = new Timeout(5 * 60 * 1000);
 
+  protected Configuration createConfiguration() {
+    Configuration conf =  super.createConfiguration();
+    //patch the read/write capacity
+    boolean scaleCapacityLimitEnabled = conf.getBoolean("fs.s3a.s3guard.ddb.table.scale.capacity.limit", true);
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/s3guard/ITestS3GuardConcurrentOps.java]: I think we should just the R/"W capacities to 0 to switch to on-demand, and remove any new configuration option. I don't care about capacity of allocation, just want to keep costs down

File: hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/testing.md [hunk 4]
```diff
@@ -854,7 +854,7 @@ and, for test runs in or near the S3/DDB stores, throttling events.
 
 If you want to manage capacity, use `s3guard set-capacity` to increase it
 (performance) or decrease it (costs).
-For remote `hadoop-aws` test runs, the read/write capacities of "10" each should suffice;
+For remote `hadoop-aws` test runs, the read/write capacities of "1" each should suffice;
```
Comment (steveloughran) [inline hadoop-tools/hadoop-aws/src/site/markdown/tools/hadoop-aws/testing.md]: Maybe review this whole paragraph to say "the default is on-demand with read/write capacities of 0" and not worry about the mention of autoscaling

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 31 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1055 | trunk passed |
| +1 | compile | 36 | trunk passed |
| +1 | checkstyle | 25 | trunk passed |
| +1 | mvnsite | 42 | trunk passed |
| +1 | shadedclient | 747 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 27 | trunk passed |
| 0 | spotbugs | 59 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 57 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 32 | the patch passed |
| +1 | compile | 30 | the patch passed |
| +1 | javac | 30 | the patch passed |
| -0 | checkstyle | 18 | hadoop-tools/hadoop-aws: The patch generated 12 new + 0 unchanged - 0 fixed = 12 total (was 0) |
| +1 | mvnsite | 34 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | xml | 2 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 750 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 24 | the patch passed |
| +1 | findbugs | 62 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 286 | hadoop-aws in the patch passed. |
| +1 | asflicense | 30 | The patch does not generate ASF License warnings. |
| | | 3376 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1037 |
| Optional Tests | dupname asflicense mvnsite compile javac javadoc mvninstall unit shadedclient findbugs checkstyle xml |
| uname | Linux 1005a8bb1108 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 1e727cf |
| Default Java | 1.8.0_212 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/2/testReport/ |
| Max. process+thread count | 446 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 32 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1080 | trunk passed |
| +1 | compile | 32 | trunk passed |
| +1 | checkstyle | 22 | trunk passed |
| +1 | mvnsite | 40 | trunk passed |
| +1 | shadedclient | 712 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 21 | trunk passed |
| 0 | spotbugs | 57 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 56 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 31 | the patch passed |
| +1 | compile | 27 | the patch passed |
| +1 | javac | 27 | the patch passed |
| +1 | checkstyle | 16 | the patch passed |
| +1 | mvnsite | 32 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | xml | 1 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 698 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 20 | the patch passed |
| +1 | findbugs | 59 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 276 | hadoop-aws in the patch passed. |
| +1 | asflicense | 28 | The patch does not generate ASF License warnings. |
| | | 3260 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=17.05.0-ce Server=17.05.0-ce base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1037 |
| Optional Tests | dupname asflicense mvnsite compile javac javadoc mvninstall unit shadedclient findbugs checkstyle xml |
| uname | Linux d38695fb56f9 4.4.0-139-generic #165-Ubuntu SMP Wed Oct 24 10:58:50 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / a2a8be1 |
| Default Java | 1.8.0_212 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/3/testReport/ |
| Max. process+thread count | 445 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
Patch LGTM

+1

when I commit I'll also cut the line "Tip: for agility, use DynamoDB autoscaling" as its obsolete now

**Comment by steveloughran** [comment]
I've committed it locally -before I push it to the ASF repo, I want to list you as the author field on the commit. What name/email addr do you want for the credit so it wires up to your github account?

(i.e. if I do a `git commit --author "name <email>"` what goes in there?

**Comment by lqjack** [comment]
got it , I will do that. thanks .

**Comment by hadoop-yetus** [comment]
:confetti_ball: **+1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 60 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| +1 | mvninstall | 1102 | trunk passed |
| +1 | compile | 35 | trunk passed |
| +1 | checkstyle | 21 | trunk passed |
| +1 | mvnsite | 37 | trunk passed |
| +1 | shadedclient | 745 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 23 | trunk passed |
| 0 | spotbugs | 69 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 67 | trunk passed |
||| _ Patch Compile Tests _ |
| +1 | mvninstall | 44 | the patch passed |
| +1 | compile | 32 | the patch passed |
| +1 | javac | 32 | the patch passed |
| +1 | checkstyle | 19 | the patch passed |
| +1 | mvnsite | 32 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | xml | 1 | The patch has no ill-formed XML file. |
| +1 | shadedclient | 764 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 21 | the patch passed |
| +1 | findbugs | 63 | the patch passed |
||| _ Other Tests _ |
| +1 | unit | 283 | hadoop-aws in the patch passed. |
| +1 | asflicense | 30 | The patch does not generate ASF License warnings. |
| | | 3479 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=18.09.8 Server=18.09.8 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/4/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1037 |
| Optional Tests | dupname asflicense mvnsite compile javac javadoc mvninstall unit shadedclient findbugs checkstyle xml |
| uname | Linux 2d82fcab1017 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 4e66cb9 |
| Default Java | 1.8.0_212 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/4/testReport/ |
| Max. process+thread count | 412 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1037/4/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

**Comment by steveloughran** [comment]
worked out your email addr from the .patch file; committed to trunk. thanks

**Comment by lqjack** [comment]
Now I found that the diff is nothing,  I could not to change the email in the commit settings.

**Comment by lqjack** [comment]
@steveloughran

**Comment by steveloughran** [comment]
the patch is merged in; closing

**Comment by steveloughran** [comment]
IF that wasn't the right addr, then add it as an email address to your gibhub a/c and you'll get the credit

**Comment by lqjack** [comment]
@steveloughran  email address : lqjacklee@gmail.com, thanks a lot.

**Pull request closed without merging**